modules/porous_flow/src/actions/PorousFlowAddMaterialJoiner.C

```diff
@@ -0,0 +1,150 @@
+
+      // If a PorousFlowJoiner material has been included in the input file,
+      // let the user know that it should be removed
+      if (material->getMooseObjectType() == "PorousFlowJoiner")
+        mooseError("PorousFlowJoiner materials are no longer required in the input file.\nPlease "
+                   "remove all PorousFlowJoiner materials from the input file and run it again");
+
+      // Add joiner material for PorousFlowSingleComponentFluid materials
+      if (material->getMooseObjectType() == "PorousFlowSingleComponentFluid")
+      {
+        // Input parameters for the material that is being added
+        InputParameters & pars = material->getObjectParams();
+        // Check if the material is evaluated at the nodes or qps
+        const bool at_nodes = pars.get<bool>("at_nodes");
+
+        // Key the addition of the joiner off the phase 0 fluid so it is only added once
+        if (pars.get<unsigned int>("phase") == 0)
+        {
+          // Join density and viscosity if they are calculated
+          if (pars.get<bool>("compute_density_and_viscosity"))
+          {
+            if (at_nodes)
+            {
+              addJoiner(
+                  at_nodes, "PorousFlow_fluid_phase_density_nodal", "PorousFlow_density_nodal_all");
+              addJoiner(at_nodes, "PorousFlow_viscosity_nodal", "PorousFlow_viscosity_nodal_all");
+            }
+            else
+            {
+              addJoiner(at_nodes, "PorousFlow_fluid_phase_density_qp", "PorousFlow_density_qp_all");
+              addJoiner(at_nodes, "PorousFlow_viscosity_qp", "PorousFlow_viscosity_qp_all");
+            }
+          }
+
+          // Join enthalpy if it is calculated
+          if (pars.get<bool>("compute_enthalpy"))
+          {
+            if (at_nodes)
+              addJoiner(at_nodes,
+                        "PorousFlow_fluid_phase_enthalpy_nodal",
+                        "PorousFlow_enthalpy_nodal_all");
+            else
+              addJoiner(
+                  at_nodes, "PorousFlow_fluid_phase_enthalpy_qp", "PorousFlow_enthalpy_qp_all");
+          }
+
+          // Join internal energy if it is calculated
+          if (pars.get<bool>("compute_internal_energy"))
+          {
+            if (at_nodes)
+              addJoiner(at_nodes,
+                        "PorousFlow_fluid_phase_internal_energy_nodal",
+                        "PorousFlow_internal_energy_nodal_all");
+            else
+              addJoiner(at_nodes,
+                        "PorousFlow_fluid_phase_internal_energy_qp",
+                        "PorousFlow_internal_energy_qp_all");
+          }
+        }
+      }
+
+      // Add joiner materials for relative permeability materials
+      // Note: this relies on the relative permeability class starting
+      // with PorousFlowRelativePermeability
+      if (material->getMooseObjectType().find("PorousFlowRelativePermeability") == 0)
```

Comment:
This bit only works if a relative permeability material is called `PorousFlowRelativePermeabilityFoo` but I couldn't think of a better way of checking for these . 

The best thing would be to see if this material inherits from `PorousFlowRelativePermeabilityBase`, but I couldn't find a way to do this